File: src/cs/Ssh/IO/SshDataReader.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.IO;
using System.Text;

namespace Microsoft.DevTunnels.Ssh.IO;

/// <summary>
/// Reads SSH-formatted data (bytes, integers, and strings) from a Buffer.
/// </summary>
public struct SshDataReader : IEquatable<SshDataReader>
{
	private readonly Buffer buffer;

	public SshDataReader(Buffer buffer)
	{
		this.buffer = buffer;
		Position = 0;
	}

	public Buffer Buffer => this.buffer;

	public int Position { get; set; }

	public long Available => this.buffer.Count - Position;

	public bool ReadBoolean()
	{
		return ReadByte() != 0;
	}

	public byte ReadByte()
	{
		if (Available == 0)
		{
			throw new EndOfStreamException();
		}

		byte value = this.buffer[Position];
		Position++;
		return value;
	}

	public uint ReadUInt32()
	{
		if (Available < 4)
		{
			throw new EndOfStreamException();
		}

		var value0 = this.buffer[Position + 0];
		var value1 = this.buffer[Position + 1];
		var value2 = this.buffer[Position + 2];
		var value3 = this.buffer[Position + 3];
		Position += 4;

		var value = ((uint)value0 << 24 | (uint)value1 << 16 | (uint)value2 << 8 | (uint)value3);
		return value;
	}

	public ulong ReadUInt64()
	{
		if (Available < 8)
		{
			throw new EndOfStreamException();
		}

		var value0 = this.buffer[Position + 0];
		var value1 = this.buffer[Position + 1];
		var value2 = this.buffer[Position + 2];
		var value3 = this.buffer[Position + 3];
		var value4 = this.buffer[Position + 4];
		var value5 = this.buffer[Position + 5];
		var value6 = this.buffer[Position + 6];
		var value7 = this.buffer[Position + 7];
		Position += 8;

		var value = (ulong)value0 << 56 | (ulong)value1 << 48 | (ulong)value2 << 40 | (ulong)value3 << 32 |
			(ulong)value4 << 24 | (ulong)value5 << 16 | (ulong)value6 << 8 | (ulong)value7;
		return value;
	}

	public string ReadString(Encoding encoding)
	{
		if (encoding == null) throw new ArgumentNullException(nameof(encoding));

		var bytes = ReadBinary();
		return encoding.GetString(bytes.Array, bytes.Offset, bytes.Count);
	}

	public string[] ReadList(Encoding encoding)
	{
		string stringList = ReadString(encoding);
		return stringList.Length == 0 ? Array.Empty<string>() : stringList.Split(',');
	}

	public BigInt ReadBigInt(bool lengthInBits = false)
	{
		uint length = ReadUInt32();
		if (lengthInBits)
		{
			length = (length + 7) / 8;
		}

		var data = ReadBinary(length);

		if (data.Count == 0)
		{
			return BigInt.Zero;
		}
		else if (lengthInBits)
		{
			return BigInt.FromByteArray(data.ToArray(), unsigned: true);
		}
		else
		{
			return new BigInt(data);
		}
	}

	public Buffer ReadBinary()
	{
		uint length = ReadUInt32();
		if (Available < length)
		{
			throw new EndOfStreamException();
		}

		return ReadBinary(length);
	}

	public Buffer ReadBinary(uint length)
	{
		var data = this.buffer.Slice(Position, (int)length);
		Position += (int)length;
		return data;
	}

	public bool Equals(SshDataReader other) => this.buffer.Equals(other.Buffer);

	public override bool Equals(object? obj) => obj is SshDataReader other && Equals(other);

	public override int GetHashCode() => this.buffer.GetHashCode();

	public static bool operator ==(SshDataReader left, SshDataReader right) => left.Equals(right);

	public static bool operator !=(SshDataReader left, SshDataReader right) => !(left == right);
}

File: src/go/ssh/io/ssh_data_reader.go
// Copyright (c) Microsoft Corporation. All rights reserved.

package sshio

import (
	"fmt"
	"math/big"
	"strings"
)

// SSHDataReader reads SSH-encoded data types from an in-memory buffer.
//
// All multi-byte integers are read in big-endian (network) byte order
// per the SSH protocol specification (RFC 4251).
type SSHDataReader struct {
	buffer   []byte
	Position int
}

// NewSSHDataReader creates a new SSHDataReader over the given buffer.
func NewSSHDataReader(buffer []byte) *SSHDataReader {
	return &SSHDataReader{buffer: buffer}
}

// Buffer returns the underlying byte buffer.
func (r *SSHDataReader) Buffer() []byte {
	return r.buffer
}

// Available returns the number of unread bytes remaining in the buffer.
func (r *SSHDataReader) Available() int {
	avail := len(r.buffer) - r.Position
	if avail < 0 {
		return 0
	}
	return avail
}

// ReadByte reads a single byte.
func (r *SSHDataReader) ReadByte() (byte, error) {
	if r.Available() < 1 {
		return 0, fmt.Errorf("attempted to read past end of buffer")
	}
	v := r.buffer[r.Position]
	r.Position++
	return v, nil
}

// ReadBoolean reads a boolean from a single byte (any non-zero value is true).
func (r *SSHDataReader) ReadBoolean() (bool, error) {
	b, err := r.ReadByte()
	if err != nil {
		return false, err
	}
	return b != 0, nil
}

// ReadUInt32 reads a 32-bit unsigned integer in big-endian byte order.
func (r *SSHDataReader) ReadUInt32() (uint32, error) {
	if r.Available() < 4 {
		return 0, fmt.Errorf("attempted to read past end of buffer")
	}
	v := uint32(r.buffer[r.Position])<<24 |
		uint32(r.buffer[r.Position+1])<<16 |
		uint32(r.buffer[r.Position+2])<<8 |
		uint32(r.buffer[r.Position+3])
	r.Position += 4
	return v, nil
}

// ReadUInt64 reads a 64-bit unsigned integer in big-endian byte order.
func (r *SSHDataReader) ReadUInt64() (uint64, error) {
	if r.Available() < 8 {
		return 0, fmt.Errorf("attempted to read past end of buffer")
	}
	v := uint64(r.buffer[r.Position])<<56 |
		uint64(r.buffer[r.Position+1])<<48 |
		uint64(r.buffer[r.Position+2])<<40 |
		uint64(r.buffer[r.Position+3])<<32 |
		uint64(r.buffer[r.Position+4])<<24 |
		uint64(r.buffer[r.Position+5])<<16 |
		uint64(r.buffer[r.Position+6])<<8 |
		uint64(r.buffer[r.Position+7])
	r.Position += 8
	return v, nil
}

// ReadString reads a UTF-8 string with a uint32 length prefix.
func (r *SSHDataReader) ReadString() (string, error) {
	data, err := r.ReadBinary()
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// ReadBinary reads a byte slice with a uint32 length prefix.
func (r *SSHDataReader) ReadBinary() ([]byte, error) {
	length, err := r.ReadUInt32()
	if err != nil {
		return nil, err
	}
	return r.ReadBytes(int(length))
}

// ReadBytes reads exactly length bytes from the buffer.
func (r *SSHDataReader) ReadBytes(length int) ([]byte, error) {
	if r.Available() < length {
		return nil, fmt.Errorf("attempted to read past end of buffer")
	}
	data := r.buffer[r.Position : r.Position+length]
	r.Position += length
	return data, nil
}

// ReadList reads a name-list: a comma-separated string with a uint32 length prefix.
// Returns an empty slice for an empty name-list.
func (r *SSHDataReader) ReadList() ([]string, error) {
	s, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	if s == "" {
		return []string{}, nil
	}
	return strings.Split(s, ","), nil
}

// ReadBigInt reads a multi-precision integer in SSH mpint format (RFC 4251).
// The value is stored as a signed two's complement big-endian integer
// with a uint32 length prefix.
func (r *SSHDataReader) ReadBigInt() (*big.Int, error) {
	data, err := r.ReadBinary()
	if err != nil {
		return nil, err
	}
	if len(data) == 0 {
		return new(big.Int), nil
	}
	return SSHBytesToBigInt(data), nil
}
